Query Refactoring: Adding getBuilderPrototype() method to all QueryParsers

cbuescher opened the pull request:
Currently there is a registry for all QueryParsers accessible via the IndicesQueriesModule. For deserializing nested queries e.g. for the BoolQueryBuilder (https://github.com/elastic/elasticsearch/pull/11121) we need to look up query builders by their name to be able to deserialize using a prototype builder of the concrete class. This PR adds a getBuilderPrototype() method to each query parser so we can re-use the parser registry to get the corresponding builder using the query name. This might change in the future when we decide to register also builders or merger parsers and builders in a later stage of the refactoring. (see https://github.com/elastic/elasticsearch/pull/11121#discussion_r30214415) for past discussion.

cbuescher committed:
Query refactoring: Adding getBuilderPrototype() method to all QueryParsers

Currently there is a registry for all QueryParsers accessible via the IndicesQueriesModule.
For deserializing nested queries e.g. for the BoolQueryBuilder (#11121) we need to look up
query builders by their name to be able to deserialize using a prototype builder of the concrete
class. This PR adds a getBuilderPrototype() method to each query parser so we can re-use the
parser registry to get the corresponding builder using the query name.

Closes #11344

javanna commented on src/main/java/org/elasticsearch/index/query/AndQueryParser.java:
```diff
@@ -36,6 +36,8 @@
 @Deprecated
 public class AndQueryParser extends BaseQueryParserTemp {
 
+    private static final AndQueryBuilder AND_QUERY_BUILDER = new AndQueryBuilder();
```
I was wondering if we should use some convention on naming, like `INSTANCE` or `PROTOTYPE`, I have no strong preference though

javanna commented on src/main/java/org/elasticsearch/index/query/FuzzyQueryBuilder.java:
```diff
@@ -62,6 +62,14 @@ public FuzzyQueryBuilder(String name, Object value) {
+     */
+    FuzzyQueryBuilder() {
+        this.name = null;
+        this.value = null;
```
In general, I don't see any other way to do this at the moment, not loving though having to assign null to so many fields...although it's only a package private constructor that does it,  maybe we could move the instance to the builders, call them prototype and make the default constructor with null assignment even private rather than package private. Or even remove it and have the INSTANCE passing in null values during object creation. Thoughts?

cbuescher replied:
I think I like the idea of moving the INSTANCE (i'd rather call them PROTOTYPE) constants from the parsers to the builders, then use existing constructors. However, then we will have to make them public (or at least package private) to access them from the parsers. If your think this is a good idea I'll go through all the classes again and change this.

javanna replied:
yea constructors would become private, which is good, you sure constants would need to become public? or is it package private enough?

cbuescher replied:
I think package private works as long as we keep parsers and builders together.

javanna replied:
sounds good, we were even thinking about merging those, we will not move them to separate packages for sure.

javanna commented on src/main/java/org/elasticsearch/index/query/LimitQueryBuilder.java:
```diff
@@ -31,12 +31,19 @@
 
     public static final String NAME = "limit";
-    private final int limit;
+    private final Integer limit;
```
not loving this change, maybe just use -1 or some other value rather than null and leave int?

javanna commented on src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java:
```diff
@@ -165,6 +165,14 @@ public MultiMatchQueryBuilder(Object text, String... fields) {
     }
 
     /**
+     * ppc
```
ppc?

cbuescher replied:
Thats strange, I think I already corrected and pushed that. Will check and redo if needed

javanna commented on src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java:
```diff
@@ -106,6 +106,13 @@ public QueryStringQueryBuilder(String queryString) {
     }
 
     /**
+     * ppc
```
maybe ppc=package private constructor? :) can you change that?

javanna commented on src/main/java/org/elasticsearch/index/query/SpanFirstQueryBuilder.java:
```diff
@@ -29,7 +29,7 @@
     private final SpanQueryBuilder matchBuilder;
 
-    private final int end;
+    private final Integer end;
```
same comment as above around using Integer rather int, would prefer the latter

javanna commented on src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java:
```diff
@@ -44,8 +44,8 @@
-    private final int offset;
-    private final int length;
+    private final Integer offset;
+    private final Integer length;
```
same here too

javanna commented:
looks good, left a few comments

cbuescher commented:
@javanna had another round, moved the constants to the builders and reverted most of the package private constructors, only introduced private constructors where needed.

javanna commented:
LGTM thanks @cbuescher

cbuescher commented:
Thanks, rebased and will merge then.

javanna commented:
I am marking this as breaking, as it breaks plugins that introduce custom queries by requiring their `QueryParser` implementation to implement the new `getPrototypeBuilder` method. The new method is supposed to return an empty builder instance corresponding to the parser. The builder will be used aas intermediate query object and serialized over the write for communication between the nodes.
